### PR TITLE
Fix helper threads giving up before more work is available

### LINKED ISSUE #2304: `HelperThread` gives up if there's no immediate work available on the pool
Honestly, I was lazy. It's a legitimate bug that I caused because I cut some corners. I need to beef up the HelperThread mechanism to sleep while waiting for new work to arrive, work which might be late, but will eventually unblock the worker threads of the pool.

Identified in https://github.com/typelevel/munit-cats-effect/pull/124. There's a quick and dirty way to avoid the deadlock, which is to let the helper threads busy wait until the fiber that unblocks the worker threads arrives. Obviously I don't need to overstate how terrible that would be.

### PR DESCRIPTION
**Pull request by vasilmkd**
This will eventually fix #2304. For now I have a reproduction as a unit test.

### CODE REVIEW

**Comment by djspiewak** [comment]
Should we target this at series/3.2.x?

File: tests/jvm/src/test/scala/cats/effect/unsafe/HelperThreadParkSpec.scala [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+      Resource.make(IO(smallRuntime()))(rt => IO(rt.shutdown())).use { rt =>
+        val io = for {
+          p <- IO(Promise[Unit]())
+          _ <- (IO.sleep(50.millis) *> IO(
+            rt.scheduler.sleep(100.millis, () => p.success(())))).start
+          _ <- IO(Await.result(p.future, Duration.Inf))
+        } yield ()
+
+        List
+          .fill(10)(io.start)
```
Comment (djspiewak) [inline tests/jvm/src/test/scala/cats/effect/unsafe/HelperThreadParkSpec.scala:67]: Do we need to increase this on machines with higher physical thread counts?

Comment (vasilmkd) [reply tests/jvm/src/test/scala/cats/effect/unsafe/HelperThreadParkSpec.scala:67]: The test runs on an artificial `IORuntime` with just 2 threads.

**Comment by vasilmkd** [comment]
I released the following snapshot based off of https://github.com/typelevel/cats-effect/pull/2305/commits/614e90f26f7046793fd2c99a9bf5c8131e5a5f81.

```scala
"io.vasilev" %%% "cats-effect" % "3.2-13-4551ba8"
```

**Comment by vasilmkd** [comment]
I did an extensive round of testing using `fs2`, which thankfully uncovered some issues while I was developing the fix. I have managed to fix those issues and the `fs2` test suite was passing reliably in the end.

**Comment by vasilmkd** [comment]
The snapshot release now makes the CI pass in https://github.com/typelevel/munit-cats-effect/pull/124.

Also, my personal fork passed each time I ran it.
https://github.com/vasilmkd/munit-cats-effect/actions/runs/1203576033

**Comment by vasilmkd** [comment]
I found another issue, working on a fix.

**Comment by vasilmkd** [comment]
Superseded by #2312.

**Pull request closed without merging**